### PR TITLE
`clamp_errors` options for Makie Errorbars and `clamp_bincounts` for histograms

### PR DESCRIPTION
**Pull request by Moelf**
This is an alternative to #116 @sfranchel 

~~I'm a bit hesitant to clip the error bars to positive~~

## ~~DO NOT MERGE~~

~~Blocked by https://github.com/MakieOrg/Makie.jl/issues/3901~~

~~The following works on older Makie but not the latest one, I think it's a bug~~

```julia
julia> h = Hist1D(;binedges=0:1, bincounts=[0.1], sumw2=[0.1])
edges: [0.0, 1.0]
bin counts: [0.1]
total count: 0.1

julia> stephist(h); errorbars!(h); current_figure()
```
image

```julia
julia> stephist(h); errorbars!(h; error_function=FHist.pearson_err); current_figure()
```

image

### CODE REVIEW

**Comment by sfranchel** [comment]
Thanks for looking into this! 

I still think it would be very useful to clip both negative/zero values and errors to positive, to do plots in log scale. I'm quite frequently getting into errors for non-compatible values for log scale. It's true that one could avoid it by choosing more wisely the binning of histograms, but when you need to run on high number of events and a bit more complex selections it isn't really obvious what's the best binning, and it's very likely to end up with empty bins. For several MC generators, it's also quite common to get negative weights, which can produce negative valued bins.

So I'd love to have an option to clip bin counts and errors! If you believe this is more of an edge case, we could set it as not being the default behaviour.

**Comment by Moelf** [comment]
Negative bin counts tell you something is happening with the MC, and users should take conscious decisions about it, we don't want to silently produce incorrect visualization is my take.

I'm happy to clamp error bars automatically

**Comment by sfranchel** [comment]
Totally agree that we shouldn't silently produce incorrect visualization, indeed I think it would bet better to not have on this feature by default.

**Comment by Moelf** [comment]
```julia
stephist(h); errorbars!(h; clamp=false); current_figure()
```
image

```julia
julia> stephist(h); errorbars!(h); current_figure()
```
image

```julia
julia> stephist(h); errorbars!(h; error_function=FHist.pearson_err); current_figure()
```
image


@sfranchel how about this?

**Comment by sfranchel** [comment]
Ok, that looks good. Maybe I'd make the default  `clamp=false`, and I'd add the same option for clamping also bin counts in barplots and scatter plots :pray:

**Comment by Moelf** [comment]
ah, right now it only clamps the error bars -- since it's a keyword argument to the `errorbars()`, I thought we want to clamp error bars above zero by default.

**Comment by Moelf** [comment]
one caveat is that until https://github.com/MakieOrg/Makie.jl/issues/3904 is addressed, `hist()` and `stephist()` won't be able to use `clamp_bincounts = true`, you need to directly use `barplot()` and `stairs()`

**Comment by Moelf** [comment]
```julia
julia> h = Hist1D(;binedges=0:2, bincounts=[-0.1, 0.1], sumw2=[0.1, 0.1])
edges: [0.0, 1.0, 2.0]
bin counts: [-0.1, 0.1]
total count: 0.0

julia> clamp_bincounts = true; clamp_errors=true;

julia> barplot(h; clamp_bincounts); errorbars!(h; clamp_bincounts, clamp_errors); current_figure()
```


| - | clamp_bincounts = false (default) | clamp_bincounts = true |
| - | -------------------------------- | -------------------------------- |
| clamp_errors = false | image | Don't do this |
| clamp_errors = true (default) | image | image |

**Comment by Moelf** [comment]
for completeness: scatter plot:

image

**Comment by sfranchel** [comment]
Thanks a lot @Moelf , this looks great!! Very much appreciated

File: ext/FHistMakieExt.jl [hunk 1]
```diff
@@ -107,37 +131,59 @@ function Makie.plot!(input::RatioHist{<:Tuple{<:Hist1D, <:Hist1D}})
...
-            if bincounts[i] - err_low[i] <= 0
-                err_low[i] = pe_low[i]
+        FHist.binerrors(error_function, h)
+    end
+    hi_errs, lo_errs = first.(errs), last.(errs)
+
+    if clamp_bincounts && clamp_errors
+        _clamp_counts_errors!(ys, lo_errs, hi_errs)
+    elseif clamp_bincounts && !clamp_errors
+        _clamp_counts(ys)
```
Comment (sfranchel) [inline ext/FHistMakieExt.jl:178]: I'm not sure if this should be allowed. Maybe it should throw a warning about possible inconsistencies on the errorbars for negative values? Or maybe it's an overkill.

Comment (Moelf) [reply ext/FHistMakieExt.jl:178]: so I think we agreed we should NOT clam

Comment (Moelf) [reply ext/FHistMakieExt.jl:178]: ah, I think in the examples I said "don't do this" but in the code I imssed it, will fix

**Pull request merged**